### PR TITLE
Parking

### PR DESCRIPTION
**Pull request by yud08**
2022 Q3
Understandable or nah

### CODE REVIEW

File: solutions/bio1/Parking/index.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { Link, SCodeBlock, SList, SCode, SText, STitle } from "components";
```
Comment (myst-6) [inline solutions/bio1/Parking/index.tsx:1]: build fails because "Link" is imported but never used

**Review by macaquedev — CHANGES_REQUESTED** [review]
Lots of issues, sorry.

File: solutions/bio1/Parking/index.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import { SCodeBlock, SCode, SText, STitle } from "components";
+
+export const Parking = () => {
+  return (
+    <>
+      <STitle>Idea</STitle>
+      <SText>
+        This type of problem (find $k-$th of something) is quite common as problem 3, and the solution to this one can be readily applied to others of its type.
+        Clearly, constructing all possible preference lists, then sorting them to find the ith one is much too slow, so you need to find a way to construct the solution letter by letter.
```
Comment (macaquedev) [inline solutions/bio1/Parking/index.tsx:9]: ith should be `$i$-th.`

File: solutions/bio1/Parking/index.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import { SCodeBlock, SCode, SText, STitle } from "components";
+
+export const Parking = () => {
+  return (
+    <>
+      <STitle>Idea</STitle>
+      <SText>
+        This type of problem (find $k-$th of something) is quite common as problem 3, and the solution to this one can be readily applied to others of its type.
+        Clearly, constructing all possible preference lists, then sorting them to find the ith one is much too slow, so you need to find a way to construct the solution letter by letter. 
+        You do this by iterating through each car (a to z), and figuring out which preference spot to give each car so that the final list is the $k-th$ lexographically, while still maintaining a valid preference list.
```
Comment (macaquedev) [inline solutions/bio1/Parking/index.tsx:10]: `$k$-th` not `$k-th$`
Also, it's spelt lexicographically not lexographically!!

Comment (macaquedev) [inline solutions/bio1/Parking/index.tsx:10]: use latex - `$a$ to $z$`

File: solutions/bio1/Parking/index.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        Taking the sample as an example, it is clear the parking slots progress like this:
+      </SText>
+      <SText>
+      <SCode>.... -{'>'} .a... -{'>'} .ab. -{'>'} cab. -{'>'} cabd</SCode>
+      </SText>
+
+      <SText>
+        So at each stage, we want to find where we can place the next car. For it to go in its correct slot, it must prefer the slot either that it ends up in, or a slot in front of that one which has no empty spaces between it and the slot which the car ends up in.
+        For example, from cab. -{'>'} cabd, d can have any preferred parking slot, as all the slots in front of it are filled, and the only empty slot is D, which is a valid preferred slot as that is the one the car d ends up in. 
+        On the other hand, if we have an arrangement like a.bc., and the next arrangement as a.bcd, D can only go in slots C, D, and E. Even though slot A is filled, there is an empty slot in between it and the E slot which car d ends up in, so it cannot be d's preferred slot.
```
Comment (macaquedev) [inline solutions/bio1/Parking/index.tsx:24]: I won't tell you every single mistake you've made, but all the ABCD's should have latex on them

File: solutions/bio1/Parking/index.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        For example, from cab. -{'>'} cabd, d can have any preferred parking slot, as all the slots in front of it are filled, and the only empty slot is D, which is a valid preferred slot as that is the one the car d ends up in. 
+        On the other hand, if we have an arrangement like a.bc., and the next arrangement as a.bcd, D can only go in slots C, D, and E. Even though slot A is filled, there is an empty slot in between it and the E slot which car d ends up in, so it cannot be d's preferred slot. 
+        So we can find all the possible preferred slots that the next letter can go in simply, but then which one do we use?
+      </SText>
+
+      <SText>
+        We want the kth preference list, so we should first keep a counter <SCode>curr</SCode> as we iterate through all the cars, and then iterate in alphabetical order through these slots, and see if we use that slot as the preferred one, then how many preference lists have the same preferences for all the currently placed cars and the considered car.
```
Comment (macaquedev) [inline solutions/bio1/Parking/index.tsx:29]: `$k$-th`

File: solutions/bio1/Parking/index.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        So we can find all the possible preferred slots that the next letter can go in simply, but then which one do we use?
+      </SText>
+
+      <SText>
+        We want the kth preference list, so we should first keep a counter <SCode>curr</SCode> as we iterate through all the cars, and then iterate in alphabetical order through these slots, and see if we use that slot as the preferred one, then how many preference lists have the same preferences for all the currently placed cars and the considered car. 
+        If that number added to <SCode>curr</SCode> is greater or equal to $k$, then we have the correct preference and we continue to the next car; otherwise, we add that to <SCode>curr</SCode>, and continue to the next possible slot. To find how many preference lists have the same preferences for all the currently placed cars and the considered one, we can observe that when all cars are placed, the number of preference lists which are the same is exactly 1.
```
Comment (macaquedev) [inline solutions/bio1/Parking/index.tsx:30]: exactly `$1$`

**Comment by macaquedev** [comment]
Also, @yud08 please insert this in sol.py, in the same directory as sol.cpp

```
from string import ascii_uppercase, ascii_lowercase
from collections import defaultdict
from functools import reduce
from operator import mul


def product(x):
    try:
        return reduce(mul, x)
    except TypeError:
        return 1

if __name__ == "__main__":
    letters = input("Please enter lowercase letters and an integer: ").strip().split()
    letters, n = letters[0], int(letters[1])
    num_cars = len(letters)
    preference_lists = []

    currently_parked = [False] * num_cars
    for current_car in range(num_cars):
        car_symbol = ascii_lowercase[current_car]

        if not any(currently_parked):
            preference_lists.append([ascii_uppercase[letters.index(car_symbol)]])
        else:
            preference_lists.append([])
            for i in range(letters.index(car_symbol), -1, -1):
                if not currently_parked[i] and i != letters.index(car_symbol):
                    break
                preference_lists[-1].append(ascii_uppercase[i])
            preference_lists[-1].reverse()
        currently_parked[letters.index(car_symbol)] = True

    solution = ""
    x = 0
    for i in range(num_cars):
        current_index = -1
        while n > 0:
            x = product([len(i) for i in preference_lists[i + 1:]])
            n -= x
            current_index += 1
        n += x
        solution += preference_lists[i][current_index]

    print(solution)
```

**Review by yud08 — COMMENTED** [review]
d

**Review by macaquedev — CHANGES_REQUESTED** [review]
You have fixed precisely *none* of my comments. Are you sure you didn't forget to push or something?

**Comment by yud08** [comment]
mb i forgot to add the index file when i committed

**Comment by myst-6** [comment]
if you're using the CLI, you can write `git add .` to add all files instead of adding them one by one

**Review by macaquedev — APPROVED** [review]

**Pull request merged**